File: codes/python/chap09/answer_A69.py
# 최대 플로용 에지의 구조체
class maxflow_edge:
	def __init__(self, to, cap, rev):
		self.to = to
		self.cap = cap
		self.rev = rev

# 깊이 우선 탐색
def dfs(pos, goal, F, G, used):
	if pos == goal:
		return F # 골 도달: 플로를 흘려 보낸다!
	# 탐색한다
	used[pos] = True
	for e in G[pos]:
		# 용량이 1 이상이고, 아직 방문하지 않은 노드에만 진행한다
		if e.cap > 0 and not used[e.to]:
			flow = dfs(e.to, goal, min(F, e.cap), G, used)
			# 플로를 흘려 보내는 경우, 잔여 그래프의 용량을 flow 만큼 증감시킨다
			if flow >= 1:
				e.cap -= flow
				G[e.to][e.rev].cap += flow
				return flow
	# 모든 에지를 탐색해도 발견하지 못했다...
	return 0

# 노드 s에서 노드 t까지의 최대 플로의 총유량을 반환한다(노드 수 N, 에지 리스트 edges)
def maxflow(N, s, t, edges):
	# 초기 상태의 잔여 그래프를 구축
	# (여기는 책과 다소 그 구현이 달라, 8번째 행은 G[a]에 추가된 뒤이므로 len(G[a]) - 1로 되어 있음에 주의)
	G = [ list() for i in range(N + 1) ]
	for a, b, c in edges:
		G[a].append(maxflow_edge(b, c, len(G[b])))
		G[b].append(maxflow_edge(a, 0, len(G[a]) - 1))
	INF = 10 ** 10
	total_flow = 0
	while True:
		used = [ False ] * (N + 1)
		F = dfs(s, t, INF, G, used)
		if F > 0:
			total_flow += F
		else:
			break # 플로를 흘려 보낼 수 없다면 조작 종류
	return total_flow

# 입력
N = int(input())
C = [ input() for i in range(N) ]

# 최대 플로를 구할 그래프를 구축한다(에지 요소는 (에지의 시작점의 번호, 에지의 종료점의 번호, 에지 용량)의 튜츨)
edges = []
for i in range(N):
	for j in range(N):
		if C[i][j] == '#':
			edges.append((i + 1, N + j + 1, 1))
for i in range(N):
	edges.append((2 * N + 1, i + 1, 1))     # 's → 파란색' 에지
	edges.append((N + i + 1, 2 * N + 2, 1)) # '빨간색 → t' 에지

# 답을 구해서 출력
answer = maxflow(2 * N + 2, 2 * N + 1, 2 * N + 2, edges)
print(answer)

File: codes/java/chap09/answer_A69.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 입력
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		String[] C = new String[N + 1];
		for (int i = 0; i < N; i++) {
			C[i + 1] = sc.next();
		}

		// 클래스를 구성한다
		MaximumFlow Z = new MaximumFlow(2 * N + 2);
		for (int i = 1; i <= N; i++) {
			for (int j = 1; j <= N; j++) {
				if (C[i].charAt(j - 1) == '#') {
					Z.addEdge(i, j + N, 1);
				}
			}
		}
		for (int i = 1; i <= N; i++) {
			Z.addEdge(2 * N + 1, i, 1);     // 's → 파란색'의 에지
			Z.addEdge(i + N, 2 * N + 2, 1); // '빨간색 → t'의 에지
		}

		// 답을 구해서 출력
		int answer = Z.maxFlow(2 * N + 1, 2 * N + 2);
		System.out.println(answer);
	}
	
	// 최대 플로를 구하기 위한 에지 의 클래스 FlowEdge
	static class FlowEdge {
		int to, cap, rev;
		public FlowEdge(int to, int cap, int rev) {
			this.to = to;
			this.cap = cap;
			this.rev = rev;
		}
	}

	// 최대 플로를 수행하는 클래스 MaximumFlow
	static class MaximumFlow {
		int n;
		boolean[] used;
		ArrayList<FlowEdge>[] G;

		// 노드 수 N인 잔여 그래프를 준비
		public MaximumFlow(int n) {
			this.n = n;
			used = new boolean[n + 1];
			G = new ArrayList[n + 1];
			for (int i = 1; i <= n; i++) {
				G[i] = new ArrayList<FlowEdge>();
			}
		}

		// 노드 a에서 b로 향하는, 상한 c 리터/초의 에지를 추가
		void addEdge(int a, int b, int c) {
			int currentGa = G[a].size();
			int currentGb = G[b].size();
			G[a].add(new FlowEdge(b, c, currentGb));
			G[b].add(new FlowEdge(a, 0, currentGa));
		}

		// 깊이 우선 탐색(F는 시작부터 pos에 도달하는 과정에서의 "잔여 그래프의 에지의 용량"의 최솟값)
		// 반환값은 흘려 보낸 플로의 양(흘려 보내지 못한 경우는 0을 반환한다)
		int dfs(int pos, int goal, int F) {
			// 골에 도착: 플로를 흘려 보낼 수 있다! 
			if (pos == goal) {
				return F;
			}
			used[pos] = true;
			// 탐색한다
			for (int i = 0; i < G[pos].size(); i++) {
				FlowEdge e = G[pos].get(i);
				// 용량 0인 에지는 사용할 수 없다
				if (e.cap == 0) {
					continue;
				}
				// 이미 방문한 노드에 가도 의미가 없다
				if (used[e.to]) {
					continue;
				}
				// 목적지까지의 패스를 찾는다
				int flow = dfs(e.to, goal, Math.min(F, e.cap));
				// 플로를 흘려 보낼 수 있는 경우, 잔여 그래프의 용량을 flow만큼만 증가시킨다
				if (flow >= 1) {
					G[pos].get(i).cap -= flow;
					G[e.to].get(e.rev).cap += flow;
					return flow;
				}
			}
			// 모든 에지를 탐색해도 찾아내지 못했다...
			return 0;
		}

		// 노드 s에서 노드 t까지의 최대 플로의 총유량을 반환한다
		int maxFlow(int s, int t) {
			final int INF = 1000000000;
			int totalFlow = 0;
			while (true) {
				Arrays.fill(used, false);
				int f = dfs(s, t, INF);
				// 플로를 흘려 보낼 수 없다면 조작 종료
				if (f == 0) {
					break;
				}
				totalFlow += f;
			}
			return totalFlow;
		}
	}
}